File: 프로그래머스/2/42860. 조이스틱/조이스틱.py
'''
1.아이디어
    - 알파벳을 아스키코드로 비교하면서 동일한 문자 나올 때까지 이동 
    - 동일 문자 나오면, 커서 이동 
    - 고려해야 할 경우의 수
    1) 위로/아래로
        대상과 "A"와의 ord() 숫자 차이 > 그만큼 이동
        이때, A>Z일때의 차이 vs A<Z일때의 차이 중 적은 것으로 선택
    2) 왼쪽으로/오른쪽으로
        각 대상의 odr()가 가장 작은 곳에서 끝나도록  ?? 우선 얘는 없이 해보자  
'''
def solution(name):
    answer = 0
    length = len(name)

    # 1. 위/아래로 바꾸는 횟수
    for ch in name:
        answer += min(ord(ch) - ord('A'), ord('Z') - ord(ch) + 1)

    # 2. 좌우로 이동하는 최소 횟수 계산
    move = length - 1  # 오른쪽으로만 가는 경우

    for i in range(length):
        next_idx = i + 1
        # 연속된 A의 끝 찾기
        while next_idx < length and name[next_idx] == 'A':
            next_idx += 1
        # 오른쪽으로 갔다가 왼쪽으로 꺾기
        move = min(move, 2 * i + length - next_idx)
        # 왼쪽으로 갔다가 오른쪽으로 꺾기
        move = min(move, i + 2 * (length - next_idx))

    answer += move
    return answer

File: 프로그래머스/2/42860. 조이스틱/조이스틱.java
/*
구해야 하는 조작횟수의 총합은
1. 구해야하는 알파벳-현재알파벳 -> 위아래로 이동해야 하는 횟수
2. A가 띄엄띄엄 있을 때는 어차피 지나가야 해서 상관 없지만,
    A가 연속으로 있을 때는 해당 구간을 지나가지 않는 것이 최소!
    이때 연속된 구간들 별로 최소인 값이 있을 수 있으므로, 연속된 구간 만날 때마다 거리 갱신
3. A가 연속인 인덱스를 구하고, 현재 위치랑 더 가까운 쪽으로 좌우로 이동하는 횟수 

*/
import java.util.*;

class Solution {
    public int solution(String name) {
        
        // 위아래 조작 필요한 알파벳 횟수 구하기
        int answer = 0;
        for(char c : name.toCharArray()){
            answer += Math.min(c-'A', 'Z'-c+1);
        }
                
        // 연속된(뛰어넘을 수 있는) A 구간 구하기
        int move = name.length() - 1;

        for(int i=0; i<name.length(); i++){

            int next = i + 1;

            while(next < name.length() && name.charAt(next) == 'A'){
                next++;
            }

            move = Math.min(move,
                    i*2 + name.length() - next); // 앞에 먼저

            move = Math.min(move,
                    i + 2*(name.length() - next)); // 뒤에 먼저
        }

        return answer+move;
    }
}